File: src/main/scala/com/celestia/factories/GLProgramBuilder.scala

package com.celestia.factories

import java.io.File
import java.util

import com.celestia.models.GLProgram
import com.celestia.interfaces.IGLProgramBuilder
import com.jogamp.opengl.GL2._ 
import com.jogamp.opengl._ 
import com.jogamp.opengl.awt.GLCanvas

/**
 * Builder class for constructing a valid GL Program
 * from multiple shader obejcts. This class can take
 * any number of shaders and will add them to the
 * private program object. After all shaders have
 * been added, the class can be told to iterate through
 * the shader collection and construct a new GLProgram
 * from the shader collection.
 */
class GLProgramBuilder(val shaders: List[(String, Int)]=List()) extends IGLProgramBuilder {

  /**
   * Add a new shader to the current program.
   */
  override def addShader(shaderPath:String, shaderType:Int):IGLProgramBuilder = new GLProgramBuilder((shaderPath, shaderType) :: shaders)


  /**
   * Construct the new program from the supplied
   * shaders.
   */
  override def build(drawable:GLAutoDrawable):GLProgram={
    // Get GL Context 
    val gl:GL4 = drawable.getGL.asInstanceOf[GL4];
    val programId:Int = gl.glCreateProgram()

    // Iterate through shader list and compile
    // all shader sources
    val shaderIds = shaders.foldLeft(List[Int]())((list, pair) => {
      val shaderSource:Array[String] = io.Source.fromFile(pair._1).map((i)=>i.toString).toArray
      val lineLengths:Array[Int] = shaderSource.map((i)=>i.length)
      val shaderId = gl.glCreateShader(pair._2)
      gl.glShaderSource(shaderId, shaderSource.length, shaderSource, lineLengths, 0)
      gl.glCompileShader(shaderId)
      gl.glAttachShader(programId, shaderId)
      shaderId :: list
    })

    // Link the program and return a new GLProgram
    // wrapper object
    gl.glLinkProgram(programId)
    shaderIds.foreach((i)=>gl.glDeleteShader(i))
    new GLProgram(programId, shaderIds)
  }
}

File: src/main/java/com/celestia/factories/GLProgramBuilder.java

package com.celestia.factories;

import com.celestia.interfaces.IGLProgramBuilder;
import com.celestia.models.GLProgram;
import com.jogamp.opengl.GL4;
import com.jogamp.opengl.GLAutoDrawable;
import graphicslib3D.GLSLUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class GLProgramBuilder implements IGLProgramBuilder {

    public static final int GL_VERTEX_SHADER = GL4.GL_VERTEX_SHADER;
    public static final int GL_FRAGMENT_SHADER = GL4.GL_FRAGMENT_SHADER;
    public static final int GL_GEOMETRY_SHDARE = GL4.GL_GEOMETRY_SHADER;
    public static final int GL_TESS_CONTROL_SHADER = GL4.GL_TESS_CONTROL_SHADER;
    public static final int GL_TESS_EVALUATION_SHADER = GL4.GL_TESS_EVALUATION_SHADER;

    // public final int programId;
    public final ArrayList<Pair<String, Integer>> shaders;

    private class Pair<A, B> {
        public final A a;
        public final B b;
        public Pair(A a, B b) {
            this.a = a;
            this.b = b;
        }
    }


    @Override 
    public IGLProgramBuilder addShader(String shaderPath, int shaderType) {
        ArrayList<Pair<String, Integer>> list = new ArrayList<Pair<String, Integer>>(shaders);
        list.add(new Pair<String, Integer>(shaderPath, shaderType));
        return new GLProgramBuilder(shaders);
    }
    
    @Override 
    public GLProgram build(GLAutoDrawable glAutoDrawable) {
        GL4 gl = (GL4) glAutoDrawable.getGL();
        GLSLUtils glslUtils = new GLSLUtils();

        int programId = gl.glCreateProgram();
        ArrayList<Integer> shaderIds = new ArrayList<Integer>();

        for(Pair<String, Integer> pair : shaders) {
            String[] shaderSource = glslUtils.readShaderSource(pair.a);
            int[] lineLengths = new int[shaderSource.length];
            for(int i=0; i < shaderSource.length; i++) {
                lineLengths[i] = shaderSource[i].length();
            }

            int shaderId = gl.glCreateShader(pair.b);
            gl.glShaderSource(shaderId, shaderSource.length, shaderSource, lineLengths, 0);
            gl.glCompileShader(shaderId);
            gl.glAttachShader(programId, shaderId);
            gl.glDeleteShader(shaderId);
            shaderIds.add(shaderId);
        }

        return new GLProgram(programId, shaderIds);
    }
    
    public GLProgramBuilder() {
        this.shaders = new ArrayList<Pair<String, Integer>>();
    }
    
    private GLProgramBuilder(ArrayList<Pair<String, Integer>> shaders) {
        this.shaders = shaders;
    }
}